dftimewolf/lib/processors/gcp_cloud_resource_tree.py

```diff
@@ -0,0 +1,1199 @@
+    output.extend(self.GenerateChildrenTree(level + 1))
+
+    return output
+
+  def GenerateChildrenTree(self, level: int) -> List[Dict[str, 'Resource']]:
+    """Generate the resource children tree.
+
+    Args:
+      level: The level in the tree to place the children at
+    Returns:
+      List of dictionaries containing a reference to the children resource and
+      their names indented based on their location in the tree
+    """
+    result: List[Dict[str, Resource]] = []
+    tab = '    '
+
+    for child in self.children:
+      entry: Dict[str, Any] = {}
+      entry['resource_object'] = child
+      entry['graph'] = f'{tab*level}|--{child.name}'
+      result.append(entry)
+
+      if child.children:
+        result.extend(child.GenerateChildrenTree(level + 1))
+
+    return result
+
+  def __str__(self) -> str:
+    """Return a string representation of the resource tree."""
+    output = '\n'
+    dashes = '-' * 200
+
+    # Draw table header
+    output = output + dashes + '\n'
+    # pylint: disable=line-too-long
+    output = output + '{:<20s}|{:<18s}|{:<20s}|{:<25s}|{:<18s}|{:<20s}|{:<25s}|{:<18s}|{:<100s}\n'.format(
+        'Resource ID', 'Resource Type', 'Creation TimeStamp', 'Created By',
+        'Creator IP Addr', 'Deletion Timestamp', 'Deleted By',
+        'Deleter IP Addr', 'Tree')
+    output = output + dashes + '\n'
+
+    result = self.GenerateTree()
+    for i in result:
+      resource: Optional[Resource] = i.get('resource_object')
+      if resource:
+        output = output + \
+            ('{:<20s}|{:<18s}|{:<20s}|{:<25s}|{:<18s}|{:<20s}|{:<25s}|{:<18s}|{:<100s}\n'.format( resource.id if ('-' not in resource.id) else "" , resource.type, resource.creation_timestamp.astimezone(timezone.utc).strftime("%Y-%m-%d %H:%M:%S") if resource.creation_timestamp else "",
+            resource.created_by, resource.creator_ip_address,  resource.deletion_timestamp.astimezone(timezone.utc).strftime("%Y-%m-%d %H:%M:%S") if resource.deletion_timestamp else "", resource.deleted_by, resource.deleter_ip_address,  i.get('graph')))
+    output = output + dashes + '\n'
+
+    return output
+
+
+class GCPCloudResourceTree(module.BaseModule):
+  """GCP Cloud Resource Tree Creator.
+
+  input: None, takes input from parameters only.
+  output: Temp file with a dump of the resource tree
+  """
+
+  def __init__(self,
+               state: DFTimewolfState,
+               name: Optional[str] = None,
+               critical: bool = True) -> None:
+    """Initialize the Cloud Resource Tree Processor.
+
+    Args:
+      state: recipe state.
+      name: The module's runtime name.
+      critical: True if the module is critical, which causes
+          the entire recipe to fail if the module encounters an error.
+    """
+    super(GCPCloudResourceTree, self).__init__(state,
+                                               name=name,
+                                               critical=critical)
+
+    self._project_id = ''
```

Comment:
same as in other comment, add type annotation for str

Reply:
Done